helper/validation/map_test.go

```diff
@@ -0,0 +1,269 @@
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("U")}),
+				},
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("UVWXYZ")}),
+				},
+			},
+		},
+		"AllGood": {
+			Value: map[string]interface{}{
+				"AB":    "12",
+				"UVWXY": "12345",
+			},
+			ExpectedDiags: nil,
+		},
+	}
+
+	fn := MapValueLenBetween(2, 5)
+
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+			diags := fn(tc.Value, cty.Path{})
+
+			checkDiagnostics(t, tn, diags, tc.ExpectedDiags)
+		})
+	}
+}
+
+func TestValidationMapKeyMatch(t *testing.T) {
+	cases := map[string]struct {
+		Value         interface{}
+		ExpectedDiags diag.Diagnostics
+	}{
+		"NoMatch": {
+			Value: map[string]interface{}{
+				"ABC":    "123",
+				"UVWXYZ": "123456",
+			},
+			ExpectedDiags: diag.Diagnostics{
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("UVWXYZ")}),
+				},
+			},
+		},
+		"AllGood": {
+			Value: map[string]interface{}{
+				"AB":    "12",
+				"UVABY": "12345",
+			},
+			ExpectedDiags: nil,
+		},
+	}
+
+	fn := MapKeyMatch(regexp.MustCompile(".*AB.*"), "")
+
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+			diags := fn(tc.Value, cty.Path{})
+
+			checkDiagnostics(t, tn, diags, tc.ExpectedDiags)
+		})
+	}
+}
+
+func TestValidationValueKeyMatch(t *testing.T) {
+	cases := map[string]struct {
+		Value         interface{}
+		ExpectedDiags diag.Diagnostics
+	}{
+		"NotStringValue": {
+			Value: map[string]interface{}{
+				"MNO":    "ABC",
+				"UVWXYZ": 123456,
+			},
+			ExpectedDiags: diag.Diagnostics{
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("UVWXYZ")}),
+				},
+			},
+		},
+		"NoMatch": {
+			Value: map[string]interface{}{
+				"MNO":    "ABC",
+				"UVWXYZ": "UVWXYZ",
+			},
+			ExpectedDiags: diag.Diagnostics{
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("UVWXYZ")}),
+				},
+			},
+		},
+		"BothBad": {
+			Value: map[string]interface{}{
+				"MNO":    "123",
+				"UVWXYZ": 123456,
+			},
+			ExpectedDiags: diag.Diagnostics{
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("MNO")}),
+				},
+				{
+					Severity:      diag.Error,
+					AttributePath: append(cty.Path{}, cty.IndexStep{Key: cty.StringVal("UVWXYZ")}),
+				},
+			},
+		},
+		"AllGood": {
+			Value: map[string]interface{}{
+				"MNO":    "ABC",
+				"UVWXYZ": "UVABY",
+			},
+			ExpectedDiags: nil,
+		},
+	}
+
+	fn := MapValueMatch(regexp.MustCompile(".*AB.*"), "")
+
+	for tn, tc := range cases {
+		t.Run(tn, func(t *testing.T) {
+			diags := fn(tc.Value, cty.Path{})
+
+			checkDiagnostics(t, tn, diags, tc.ExpectedDiags)
+		})
+	}
+}
+
+func checkDiagnostics(t *testing.T, tn string, got, expected diag.Diagnostics) {
```

Comment:
I've added this for now. As you mention, maybe replace with a method on `Diagnostics`?

Reply:
This is good enough for now